pkg/apiserver/clusterinfo/host.go

```diff
@@ -0,0 +1,354 @@
+	Free   int    `json:"free"`
+	Total  int    `json:"total"`
+}
+
+type HostInfo struct {
+	IP         string `json:"ip"`
+	CPUCore    int    `json:"cpu_core"`
+	CPUUsage   `json:"cpu_usage"`
+	Memory     `json:"memory"`
+	Partitions []PartitionInstance `json:"partitions"`
+}
+
+type Instance struct {
+	Address    string `json:"address"`
+	ServerType string `json:"server_type"`
+}
+
+type PartitionInstance struct {
+	Partition `json:"partition"`
+	Instance  `json:"instance"`
+}
+
+func GetAllHostInfo(db *gorm.DB) ([]HostInfo, error) {
+	hostMap, err := loadHosts(db)
+	if err != nil {
+		return nil, err
+	}
+	cores, err := loadCPUCores(db)
+	if err != nil {
+		return nil, err
+	}
+	usages, err := loadCPUUsage(db)
+	if err != nil {
+		return nil, err
+	}
+	memory, err := loadMemory(db)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := make([]HostInfo, 0)
+	for ip, instances := range hostMap {
+		var disks = make([]PartitionInstance, 0)
+		for _, instance := range instances {
+			dataDir, err := queryDeployInfo(db, instance)
+			if err != nil {
+				continue
+			}
+			diskMap, err := queryPartition(db, instance)
+			if err != nil {
+				continue
+			}
+			disk := inferPartition(dataDir, diskMap)
+
+			disks = append(disks, PartitionInstance{
+				Partition: disk,
+				Instance:  instance,
+			})
+		}
+
+		info := HostInfo{
+			IP:         ip,
+			CPUCore:    cores[ip],
+			CPUUsage:   usages[ip],
+			Memory:     memory[ip],
+			Partitions: disks,
+		}
+		infos = append(infos, info)
+	}
+
+	return infos, nil
+}
+
+func splitPath(path string) []string {
+	if path == "" {
+		return []string{}
+	}
+	return strings.Split(path, string(filepath.Separator))
+}
+
+func inferPartition(dataDir string, diskMap PartitionMap) Partition {
+	var targetDisk Partition
+	var minRelativePathLength = math.MaxInt64
+
+	for _, disk := range diskMap {
+		rel, err := filepath.Rel(disk.Path, dataDir)
+		if err != nil {
+			continue
+		}
+		var relativePathLength int
+		for _, dir := range splitPath(rel) {
+			if dir == ".." {
+				relativePathLength = -1
+				break
+			} else {
+				relativePathLength++
+			}
+		}
+		if relativePathLength == -1 {
+			continue
+		}
+		if relativePathLength < minRelativePathLength {
+			minRelativePathLength = relativePathLength
+			targetDisk = disk
+		}
+	}
+
+	return targetDisk
+}
+
+type HostMap map[string][]Instance
+
+func loadHosts(db *gorm.DB) (HostMap, error) {
+	hostMap := make(HostMap)
+
+	sql := "select TYPE, INSTANCE from INFORMATION_SCHEMA.CLUSTER_INFO;"
+	rows, err := db.Raw(sql).Rows()
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var instance, serverType string
+		err = rows.Scan(&serverType, &instance)
+		if err != nil {
+			continue
+		}
+		ip := parseIP(instance)
+		var list []Instance
+		if instances, ok := hostMap[ip]; ok {
+			list = instances
+		}
+
+		list = append(list, Instance{
+			Address:    instance,
+			ServerType: serverType,
+		})
+		hostMap[ip] = list
+	}
```

Comment:
Maybe we should load this table once, then parse it? Otherwise I guess full hardware info is loaded multiple times.

```go
type clusterInfoTableModel struct {
    Type ...
    Instance ...
    DeviceType ...
    DeviceName ...
    Name ...
    Value ...
}

var records []clusterInfoTableModel

db.Table('INFORMATION_SCHEMA.CLUSTER_INFO').Find(&records)

for ... {
    switch rec.DeviceType {
        ....
    }
}
```